cluster-essentials/pkg/client/install.go

```diff
@@ -0,0 +1,109 @@
+const defaultTimeout = 15 * time.Minute
+
+var clusterEssentialPackages = []string{"kapp-controller", "secretgen-controller"}
+
+func validatePreInstall(restConfig *rest.Config) error {
+	clientset, _ := kubernetes.NewForConfig(restConfig)
+	nsList, _ := clientset.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
+	for _, n := range nsList.Items {
+		for _, packageName := range clusterEssentialPackages {
+			deploymentsClient := clientset.AppsV1().Deployments(n.Name)
+			result, getErr := deploymentsClient.Get(context.TODO(), packageName, metav1.GetOptions{})
+			if getErr != nil {
+				//TODO check for CRD if deployment is not present
+				// if Deplyoment and CRD is not present, continue the loop to check in other namespace 
+				continue
+			}
+			// If deployment is present, check for annotation
+			_, ok := result.Annotations["tanzu.vmware.cluster.essentia/package-name"]
+			//If annotation is not present then return error
+			if !ok {
+				return fmt.Errorf("Pre validation check failed for package %s, package is not a part of cluster essential", packageName)
+			}
+	                //If above annotation is present, get the annotation for cluster essential version
+			// In case of downgrade - Exit
+			// In case of upgrade - move forward with installation
+		}
+	}
+        return nil
+}
+
+func validatePostInstall(restConfig *rest.Config, timeout time.Duration) error {
+        client, _ := kubernetes.NewForConfig(restConfig)
+        for _, packageName := range clusterEssentialPackages {
+                label := "app="+packageName
+                pods,_ := client.CoreV1().Pods(packageName).List(context.TODO(),
+                     metav1.ListOptions{LabelSelector: label})
+                if err := wait.Poll(time.Second*2, time.Second*20, func() (done bool, err error) {
+                        pod, _ := client.CoreV1().Pods(packageName).Get(context.Background(), pods.Items[0].Name, metav1.GetOptions{})
+                        if err != nil {
+                                return false, err
+                        }
+                        if pod.Status.Phase != "Running" {
+                                return false, nil
+                        }
+                        return true, nil
+                });err != nil {
+                        return err
+                }
+        }
+	return nil
+}
+
+// Install installs Cluster Essentials packages.
+// This function attempts to install packages from the given imgpkg bundle path
+// and waits until install is complete or the given timeout is reached, whichever
+// occurs first.
+func Install(ctx context.Context, config *rest.Config, imgpkgBundlePath string, timeout time.Duration) error {
+	if timeout == time.Duration(0) {
+		timeout = defaultTimeout
+	}
+	bundleDir, err := os.MkdirTemp(defaultBundleDir, "cluster-essential")
+	defer os.RemoveAll(bundleDir)
+	if err != nil {
+		return err
+	}
+	err = downloadBundles(imgpkgBundlePath, bundleDir)
```

Comment:
can we re-order download and preinstall so that we don't need to download if pre-install fails.

Reply:
We need bundle beforehand to find out version